snuba/query/parser/validation/__init__.py

```diff
@@ -37,3 +43,51 @@ def validate_query(query: Query) -> None:
     for exp in query.get_all_expressions():
         for v in validators:
             v.validate(exp, query.get_from_clause())
+
+
+class QueryValidator(ABC):
+    """
+    Validates logic that requires multiple parts of a query.
+    """
+
+    @abstractmethod
+    def validate(
+        self, entity: Entity, query: Query, alias: Optional[str] = None
+    ) -> None:
+        """
+        For the given entity, validate that the query is correct. If the query
+        is not valid, raise an Exception, otherwise return None.
+
+        :param entity: The entity to base the validation on,
+        :type entity: Entity
+        :param query: The query to validate.
+        :type query: Query
+        :param alias: The alias of the entity in a JOIN query.
+        :type alias: Optional[str]
+        :raises InvalidQueryException: [description]
+        """
+        raise NotImplementedError
+
+
+from snuba.query.parser.validation.conditions import EntityRequiredColumnValidator
+
+query_validators = [EntityRequiredColumnValidator()]
+
+
+def validate_entities_with_query(query: Query) -> None:
+    if isinstance(query, LogicalQuery):
+        entity = get_entity(query.get_from_clause().key)
+        for v in query_validators:
+            v.validate(entity, query)
+    else:
+        from_clause = query.get_from_clause()
+        if isinstance(from_clause, (LogicalQuery, CompositeQuery)):
+            return validate_entities_with_query(from_clause)
+
+        assert isinstance(from_clause, JoinClause)  # mypy
+        alias_map = from_clause.get_alias_node_map()
+        for alias, node in alias_map.items():
+            assert isinstance(node.data_source, QueryEntity)  # mypy
```

Comment:
No such luck. I think it's because the from_clause is `Union[Entity, ProcessableQuery[Entity]]` and the earlier if doesn't match the ProcessableQuery exactly.